### PR TITLE
dlpack: Update to DLPack v1.0 and Python array API v2023.12

### PR DESCRIPTION
**Pull request by dalcinl**
#### References
* [dlpack](https://dmlc.github.io/dlpack/latest/python_spec.html#python-spec)
* [array-apis](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html#array_api.array.__dlpack__)

### CODE REVIEW

File: src/mpi4py/MPI.src/asdlpack.pxi [hunk 1]
```diff
@@ -189,17 +205,22 @@ cdef int Py_CheckDLPackBuffer(object obj) noexcept:
     except: return 0  # ~> uncovered  # noqa
 
 cdef int Py_GetDLPackBuffer(object obj, Py_buffer *view, int flags) except -1:
+    cdef int version = 1
+    cdef const char *capsulename = b"dltensor_versioned"
```
Comment (dalcinl) [inline src/mpi4py/MPI.src/asdlpack.pxi:209]: @leofang Please confirm the capsule names corresponding to v1.0 need a "_versioned" prefix.

**Comment by dalcinl** [comment]
@leofang Are you aware of any other consumer implementation for v1.0, or am I the first to sail this boat?

File: src/mpi4py/MPI.src/asdlpack.pxi [hunk 2]
```diff
@@ -208,40 +231,61 @@ cdef int Py_GetDLPackBuffer(object obj, Py_buffer *view, int flags) except -1:
...
+        if device_type == kDLCPU:
+            capsule = dlpack()
+        else:
+            capsule = dlpack(stream=-1)
+
+    if not PyCapsule_IsValid(capsule, capsulename):
+        raise BufferError("dlpack: invalid capsule object")
+    pointer = PyCapsule_GetPointer(capsule, capsulename)
+    if version >= 1:
+        managed1 = <DLManagedTensorVersioned*> pointer
```
Comment (dalcinl) [inline src/mpi4py/MPI.src/asdlpack.pxi:254]: @leofang Should we add a sanity check to make sure that the (major) version value is the expected one (>=1), or should we just trust the producer?

Comment (leofang) [reply src/mpi4py/MPI.src/asdlpack.pxi:254]: What kind of sanity check do you have in mind? I don't see how we can avoid a type cast here?

**Comment by leofang** [comment]
Sorry for delay, looks good to me for a quick read. Do you want to merge this now or wait for NumPy/CuPy to catch up?

File: src/mpi4py/MPI.src/asdlpack.pxi (hunk 2 above)
Comment (dalcinl) [reply src/mpi4py/MPI.src/asdlpack.pxi:254]: I meant just something like the following assert-based check:

```python
if version >= 1:
        managed1 = <DLManagedTensorVersioned*> pointer
        assert managed1.version.major >= 1
```

**Comment by dalcinl** [comment]
>  Do you want to merge this now or wait for NumPy/CuPy to catch up?

I'm really thinking on merging it and finally making new release as soon as possible. Do you agree?

File: src/mpi4py/MPI.src/asdlpack.pxi (hunk 2 above)
Comment (leofang) [reply src/mpi4py/MPI.src/asdlpack.pxi:254]: Doesn't hurt to add if you like.

**Comment by leofang** [comment]
Sure, fine with me. I need to squeeze time to implement it for CuPy...

**Pull request merged**